### PR TITLE
Add msix and appx support to win_package

### PR DESCRIPTION
**Pull request by agowa**
##### SUMMARY

Add msix and appx support to win_package

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

win_package

##### ADDITIONAL INFORMATION

Implements #50765

### CODE REVIEW

**Comment by ahmadgb** [comment]
a check failed, i don't think i can verify

**Comment by dagwieers** [comment]
Sure you can, just click on the Details-link and scroll through the failure-logs.
```
00:20 Run command: test/sanity/pslint/pslint.ps1 lib/ansible/modules/windows/win_package.ps1
00:22 ERROR: Command "test/sanity/pslint/pslint.ps1 lib/ansible/modules/windows/win_package.ps1" returned exit status 1.
00:22 >>> Standard Error
00:22 Invoke-ScriptAnalyzer : Parse error in file /root/ansible/lib/ansible/modules/windows/win_package.ps1:  You must provide a value expression following the '-or' operator at line 174 column 93.
00:22 At /root/ansible/test/sanity/pslint/pslint.ps1:16 char:25
00:22 + ... $Results += Invoke-ScriptAnalyzer -Path $Path -Setting $PSScriptRoot/ ...
00:22 +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
00:22 + CategoryInfo          : ParserError: (ExpectedValueExpression:String) [Invoke-ScriptAnalyzer], ParseException
00:22 + FullyQualifiedErrorId : Parse error in file /root/ansible/lib/ansible/modules/windows/win_package.ps1:  You must provide a value expression following the '-or' operator at line 174 column 93.,Microsoft.Windows.PowerShell.ScriptAnalyzer.Commands.InvokeScriptAnalyzerCommand
```

Syntax error on line 174: https://github.com/ansible/ansible/pull/50864/files#diff-015c6ab28b2a53b29b2fa4d1d6c225ffR174

File: lib/ansible/modules/windows/win_package.ps1 [hunk 1]
```diff
@@ -164,6 +169,11 @@ Function Get-ProgramMetadata($state, $path, $product_id, $credential, $creates_p
     if ($path -ne $null) {
         if ($path.EndsWith(".msi", [System.StringComparison]::CurrentCultureIgnoreCase)) {
             $metadata.msi = $true
+        } elseif(
+            $path.EndsWith(".msix", [System.StringComparison]::CurrentcultureIgnoreCase) -or 
+            $path.EndsWith(".appx", [System.StringComparison]::CurrentcultureIgnoreCase) -or
```
Comment (ahmadgb) [inline lib/ansible/modules/windows/win_package.ps1]: fix syntax please

**Review by ahmadgb — APPROVED** [review]

**Comment by varunchopra** [comment]
I think tests for this would be nice.

Also, I think `appx`, `dism` and `PackageManagement` shouldn't be hard requirements to run this module.

**Comment by agowa** [comment]
dism is a hard requirement, as the command Add-AppxProvisionedPackage belongs to it, without it there is no possibility to instal appx/msix for all users.
appx is currently not used, I forgot to remove it before commiting, it is required for installing appx/msix for a single user
PackageManagement is required for Get-Package to get a list of installed software. This is the only way to check if it is already installed or not (I'm not quite sure if it uses appx underneath for querying the appx/msix-es though).

Or do you just mean to replace the `#Requires foo` with something like `import-module -name foo -erroraction stop`?

**Comment by varunchopra** [comment]
That is correct.

Check if the user is installing an appx/msix. Then check if the required modules are present.

**Comment by agowa** [comment]
@ahmadgb can you please verify this version, too? I've currently no Windows Box to test this.

**Comment by jborean93** [comment]
I'm starting to think `win_package` is being overloaded with too many things to do. I'm wondering whether we should have a separate module to install UWAs instead of trying to shoehorn it into here. I'll add it to the discuss at the next Windows Working Group. It would be great if you could come along https://github.com/ansible/community/issues/420#issuecomment-456277586.

File: lib/ansible/modules/windows/win_package.ps1 [hunk 2]
```diff
@@ -25,6 +25,7 @@ $validate_certs = Get-AnsibleParam -obj $params -name "validate_certs" -type "bo
 $creates_path = Get-AnsibleParam -obj $params -name "creates_path" -type "path"
 $creates_version = Get-AnsibleParam -obj $params -name "creates_version" -type "str"
 $creates_service = Get-AnsibleParam -obj $params -name "creates_service" -type "str"
+$isAllUsers = Get-AnsibleParam -obj $params -name "AllUsers" -type "bool" -default $true -aliases "isAllUser"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_package.ps1]: When creating module options it's best to keep them align with the Ansible standard of snake_case, e.g. `install_all_users` or something like that. It is also best to have the variable reflect the same format of the actual option and keep it aligned to the existing standard in the module.

**Review by jborean93 — CHANGES_REQUESTED** [review]

**Comment by agowa338** [comment]
@jborean93 can you update you review?

**Comment by jborean93** [comment]
@agowa338, my last comment is still stands, I’m fine with the code but this still need some integration tests.

**Comment by jborean93** [comment]
This functionality has been added with https://github.com/ansible/ansible/pull/66931. Not everything in this PR was added there but now win_package can do the basic install and uninstall of appx/msix packages.

If more functionality is required it should be done in a separate module, similar to how `package` is generic for all package managers but more features can be done in ones like `dnf`, `apt`, and so on.

**Pull request closed without merging**